src/imports/background/import-item-creation.js

```diff
@@ -0,0 +1,195 @@
+import chunk from 'lodash/fp/chunk'
+
+import db from 'src/pouchdb'
+import encodeUrl from 'src/util/encode-url-for-id'
+import { pageKeyPrefix, convertPageDocId } from 'src/page-storage'
+import { bookmarkKeyPrefix, convertBookmarkDocId } from 'src/bookmarks'
+import { checkWithBlacklist } from 'src/blacklist'
+import { isLoggable } from 'src/activity-logger'
+import { differMaps } from 'src/util/map-set-helpers'
+import {
+    IMPORT_TYPE,
+    OLD_EXT_KEYS,
+    DEF_ALLOW,
+} from 'src/options/imports/constants'
+
+// Binds old ext bookmark urls to a function that transforms old ext data to an import item
+const transformOldExtDataToImportItem = bookmarkUrls => (item, index) => ({
+    type: IMPORT_TYPE.OLD,
+    url: item.url,
+    timestamp: item.time,
+    hasBookmark: bookmarkUrls.has(item.url),
+    index,
+})
+
+// Binds an import type to a function that transforms a history/bookmark doc to an import item.
+const transformBrowserItemToImportItem = type => item => ({
+    browserId: item.id,
+    url: item.url,
+    // HistoryItems contain lastVisitTime while BookmarkTreeNodes contain dateAdded
+    timestamp: item.lastVisitTime || item.dateAdded,
+    type,
+})
+
+async function checkWithExistingBookmarks() {
+    const { rows: existingBmDocs } = await db.allDocs({
+        startkey: bookmarkKeyPrefix,
+        endkey: `${bookmarkKeyPrefix}\uffff`,
+    })
+
+    const existingIds = new Set(
+        existingBmDocs.map(row => convertBookmarkDocId(row.id).url),
+    )
+    return url => existingIds.has(url)
+}
+
+/**
+ * @returns {({ url: string }) => boolean} A function affording checking of a URL against the
+ *  URLs of existing page docs.
+ */
+async function checkWithExistingPages() {
+    const { rows: existingPageDocs } = await db.allDocs({
+        startkey: pageKeyPrefix,
+        endkey: `${pageKeyPrefix}\uffff`,
```

Comment:
what is uffff?

Reply:
This is the highest 2-byte unicode character ￿. Basically it's saying "all docs with key prefix `page/`"
Used in Pouch range queries like this as Pouch keeps things sorted internally by doc IDs, and this character will have a higher sort order than anything else we store. Maybe some more info in the [Pouch docs here](https://pouchdb.com/api.html#batch_fetch) (they actually use the character `\ufff0` here. Could you think of any difference in behaviour?)

Reply:
probably not, I was just curious